
**Comentario del Revisor**

Hola!

Soy Juan Manuel Romero, pero siéntete libre de llamarme Juanma. Soy code reviewer en Tripleten y hoy estaré revisando tu entrega.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión. 

Solo un aviso rápido: cuando estés revisando el proyecto, por favor deja mis comentarios originales tal como están. De esta manera, podemos seguir fácilmente el progreso y asegurarnos de que no se nos pase nada por alto. Y, si realizas algún cambio basado en mis comentarios, sería genial si pudieras resaltar esas actualizaciones para que se destaquen.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como estas:

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Éxito. Todo se ha hecho correctamente.

</div>


<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.

</div> 


<div class="alert alert-block alert-danger">
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Requiere corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con los comentarios en rojo.

</div>

Puedes responderme usando esto:

<div class="alert alert-block alert-info"> <b>Respuesta del estudiante.</b> <a class="tocSkip"></a> </div>

In [ ]:

<div class="alert alert-block alert-success"> 
<b>COMENTARIO GENERAL #1</b> <a class="tocSkip"></a>

Felicidades Ferc! Tu trabajo es excelente. 

Has completado todos los items necesarios para aprobar la entrega. Sigue así!

</div>



Como parte del equipo de análisis, lo primero que debes hacer es evaluar la calidad de una muestra de datos recopilados y prepararla para analizarla posteriormente. Después, en la segunda parte de este proyecto en el segundo sprint, desarrollarás más tus habilidades y harás tu primer análisis completo, respondiendo a las necesidades del cliente.

Estos son los datos que el cliente nos proporcionó. Tienen el formato de una lista de Python, con las siguientes columnas de datos:

- **user_id:** Identificador único para cada usuario.
- **user_name:** El nombre del usuario.
- **user_age:** La edad del usuario.
- **fav_categories:** Categorías favoritas de los artículos que compró el usuario, como 'ELECTRONICS', 'SPORT' y 'BOOKS' (ELECTRÓNICOS, DEPORTES y LIBROS), etc.
- **total_spendings:** Una lista de números enteros que indican la cantidad total gastada en cada una de las categorías favoritas.


In [ ]:
users = [
    ['32415', ' mike_reed ', 32.0, ['ELECTRONICS', 'SPORT', 'BOOKS'], [894, 213, 173]],
    ['31980', 'kate morgan', 24.0, ['CLOTHES', 'BOOKS'], [439, 390]],
    ['32156', ' john doe ', 37.0, ['ELECTRONICS', 'HOME', 'FOOD'], [459, 120, 99]],
    ['32761', 'SAMANTHA SMITH', 29.0, ['CLOTHES', 'ELECTRONICS', 'BEAUTY'], [299, 679, 85]],
    ['32984', 'David White', 41.0, ['BOOKS', 'HOME', 'SPORT'], [234, 329, 243]],
    ['33001', 'emily brown', 26.0, ['BEAUTY', 'HOME', 'FOOD'], [213, 659, 79]],
    ['33767', ' Maria Garcia', 33.0, ['CLOTHES', 'FOOD', 'BEAUTY'], [499, 189, 63]],
    ['33912', 'JOSE MARTINEZ', 22.0, ['SPORT', 'ELECTRONICS', 'HOME'], [259, 549, 109]],
    ['34009', 'lisa wilson ', 35.0, ['HOME', 'BOOKS', 'CLOTHES'], [329, 189, 329]],
    ['34278', 'James Lee', 28.0, ['BEAUTY', 'CLOTHES', 'ELECTRONICS'], [189, 299, 579]],
]


# Paso 1

Store 1 tiene como objetivo garantizar la coherencia en la recopilación de datos. Como parte de esta iniciativa, se debe evaluar la calidad de los datos recopilados sobre los usuarios y las usuarias. Te han pedido que revises los datos recopilados y propongas cambios. A continuación verás datos sobre un usuario o una usuaria en particular; revisa los datos e identifica cualquier posible problema.


In [36]:
user_id = '32415' #el valor esta en string y no en int
user_name = ' mike_reed ' # tiene espacios al inicio y al final y un guion en medio del nombre
user_age = 32.0 #esta en float y no en int
fav_categories = ['ELECTRONICS', 'SPORT', 'BOOKS'] #todo esta en yusculas 

[32415, 'mike reed', 32.0, ['electronics', 'sport', 'books']]


**Opciones:**

1. Se debe cambiar el tipo de dato de `user_id` de cadena a entero.
    
2. La variable `user_name` contiene una cadena que tiene espacios innecesarios y un guion bajo entre el nombre y el apellido.
    
3. El tipo de dato de `user_age` es correcto y no hay necesidad de convertirlo.
    
4. La lista `fav_categories` contiene cadenas en mayúsculas. En lugar de ello, debemos convertir los valores de la lista en minúsculas.


#Escribe tu respuesta y explica tu argumentación:

#el valor esta en string y no en int
# tiene espacios al inicio y al final y un guion en medio del nombre
#esta en float y no en int
#todo esta en mayusculas

Para cada una de las opciones, escribe en la siguiente celda markdown si la identificaste como un problema real en los datos o no. Justifica tu razonamiento. Por ejemplo, si crees que la primera opción es correcta, escríbelo y explica por qué piensas que es correcta.

<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Muy buenas respuestas! 

</div>

Para complementar, ten en cuenta lo siguiente:

- **Índice como entero o cadena**:  
    La elección entre usar un entero o una cadena para representar el índice depende del propósito de la variable y cómo planeamos usarla. Si el índice se utilizará para realizar cálculos o como referencia numérica en operaciones lógicas o matemáticas, lo más adecuado es definirlo como un entero (`int`). Por otro lado, si el índice se trata como una etiqueta, como en identificadores de usuario o códigos alfanuméricos (por ejemplo, "A123"), es mejor usar una cadena (`str`). Elegir correctamente el tipo de dato no solo mejora la claridad del código, sino también su eficiencia y mantenibilidad.
    
- **Formato de `user_name`**:  
    Es importante almacenar la información en un formato que facilite su manejo. Si `user_name` contiene espacios innecesarios o caracteres adicionales como guiones bajos entre el nombre y el apellido, puede dificultar operaciones comunes como búsquedas, comparaciones o visualizaciones.
    
- **Edad como entero**:  
    La edad, por su naturaleza, es siempre un número entero (`int`), ya que representa una cantidad discreta (años completos). Al almacenarla como un número entero en lugar de un flotante (`float`), no solo ahorramos memoria, sino que también simplificamos validaciones y operaciones. Además, usar un entero refuerza la semántica de los datos, lo que mejora la comprensión del código.
    
- **Uso consistente de mayúsculas o minúsculas**:  
    Normalizar el formato de texto (todo en mayúsculas o todo en minúsculas) es una práctica común para facilitar operaciones como búsquedas, comparaciones o clasificaciones. Por ejemplo, al comparar dos cadenas, los formatos inconsistentes pueden llevar a resultados inesperados. Ya depende de la naturaleza de los datos si tiene sentido o no utilizar mayúsculas o minúsculas. 

</div>

# Paso 2

Vamos a implementar los cambios que identificamos. Primero, necesitamos corregir los problemas de la variable `user_name` Como vimos, tiene espacios innecesarios y un guion bajo como separador entre el nombre y el apellido; tu objetivo es eliminar los espacios y luego reemplazar el guion bajo con el espacio.


In [39]:
user_name = ' mike_reed '
user_name = user_name.strip()
user_name = user_name.replace("_"," ")

print(user_name)

mike reed


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has utilizado _strip_ y _replace_ de manera correcta. Estos métodos son muy útiles para limpiar y ajustar cadenas de texto, eliminando espacios innecesarios y reemplazando caracteres no deseados. En este caso, lograste obtener `mike reed`, justo lo que esperábamos.

</div>

********Hint********

Existe un método, `strip()`, que puede eliminar espacios al principio y al final de una cadena. Además, el método `replace()` se puede usar para reemplazar una parte de una cadena. En este caso, queremos reemplazar los guiones bajos (`_`) con espacios.


# Paso 3

Luego, debemos dividir el `user_name` (nombre de usuario o usuaria) actualizado en dos subcadenas para obtener una lista que contenga dos valores: la cadena para el nombre y la cadena para el apellido.

In [43]:
user_name = 'mike reed'
name_split = user_name.split()

print(name_split)

['mike', 'reed']


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Muy bien! Utilizaste el método _split_ correctamente, dividiendo la cadena en `['mike', 'reed']`.

</div>

********Hint********

El método `split()` se utiliza para dividir una cadena. Por defecto, utiliza un espacio como separador.


# Paso 4

¡Genial! Ahora debemos trabajar con la variable `user_age`. Como ya mencionamos, esta tiene un tipo de datos incorrecto. Arreglemos este problema transformando el tipo de datos y mostrando el resultado final.


In [44]:
user_age = 32.0
user_age = int(user_age)

print(user_age)

32


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has convertido correctamente la cadena a un número entero con _int_, obteniendo `32`.

</div>

********Hint********

¿Qué tipo de datos eliminará la parte de coma flotante?


# Paso 5

Como sabemos, los datos no siempre son perfectos. Debemos considerar escenarios en los que el valor de `user_age` no se pueda convertir en un número entero. Para evitar que nuestro sistema se bloquee, debemos tomar medidas con anticipación.

Escribe un código que intente convertir la variable `user_age` en un número entero y asigna el valor transformado a `user_age_int`. Si el intento falla, mostramos un mensaje pidiendo al usuario o la usuaria que proporcione su edad como un valor numérico con el mensaje: `Please provide your age as a numerical value.` (Proporcione su edad como un valor numérico.)

In [46]:
user_age = 'treinta y dos'

try:
    
   user_age_int = int(user_age)
   print(user_age_int)
except:
    print("Please provide your age as a numerical value.")

Please provide your age as a numerical value.


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has implementado _try_ y _except_ correctamente para manejar errores. Este enfoque es muy útil para evitar que tu programa se detenga inesperadamente por errores no controlados.

</div>

********Hint********

Utiliza un bloque `try-except` para intentar la conversión; si falla, proporciona un mensaje claro indicando que la entrada debe ser numérica.

# Paso 6

El equipo de dirección de Store 1 te pidió ayudarles a organizar los datos de sus clientes para analizarlos y gestionarlos mejor.

Tu tarea es ordenar esta lista por ID de usuario de forma ascendente para que sea más fácil acceder a ella y analizarla.


In [47]:
users = [
    ['32415', ' mike_reed ', 32.0, ['ELECTRONICS', 'SPORT', 'BOOKS'], [894, 213, 173]],
    ['31980', 'kate morgan', 24.0, ['CLOTHES', 'BOOKS'], [439, 390]],
    ['32156', ' john doe ', 37.0, ['ELECTRONICS', 'HOME', 'FOOD'], [459, 120, 99]],
    ['32761', 'SAMANTHA SMITH', 29.0, ['CLOTHES', 'ELECTRONICS', 'BEAUTY'], [299, 679, 85]],
    ['32984', 'David White', 41.0, ['BOOKS', 'HOME', 'SPORT'], [234, 329, 243]],
    ['33001', 'emily brown', 26.0, ['BEAUTY', 'HOME', 'FOOD'], [213, 659, 79]],
    ['33767', ' Maria Garcia', 33.0, ['CLOTHES', 'FOOD', 'BEAUTY'], [499, 189, 63]],
    ['33912', 'JOSE MARTINEZ', 22.0, ['SPORT', 'ELECTRONICS', 'HOME'], [259, 549, 109]],
    ['34009', 'lisa wilson ', 35.0, ['HOME', 'BOOKS', 'CLOTHES'], [329, 189, 329]],
    ['34278', 'James Lee', 28.0, ['BEAUTY', 'CLOTHES', 'ELECTRONICS'], [189, 299, 579]],
]

users.sort()

print(users)

[['31980', 'kate morgan', 24.0, ['CLOTHES', 'BOOKS'], [439, 390]], ['32156', ' john doe ', 37.0, ['ELECTRONICS', 'HOME', 'FOOD'], [459, 120, 99]], ['32415', ' mike_reed ', 32.0, ['ELECTRONICS', 'SPORT', 'BOOKS'], [894, 213, 173]], ['32761', 'SAMANTHA SMITH', 29.0, ['CLOTHES', 'ELECTRONICS', 'BEAUTY'], [299, 679, 85]], ['32984', 'David White', 41.0, ['BOOKS', 'HOME', 'SPORT'], [234, 329, 243]], ['33001', 'emily brown', 26.0, ['BEAUTY', 'HOME', 'FOOD'], [213, 659, 79]], ['33767', ' Maria Garcia', 33.0, ['CLOTHES', 'FOOD', 'BEAUTY'], [499, 189, 63]], ['33912', 'JOSE MARTINEZ', 22.0, ['SPORT', 'ELECTRONICS', 'HOME'], [259, 549, 109]], ['34009', 'lisa wilson ', 35.0, ['HOME', 'BOOKS', 'CLOTHES'], [329, 189, 329]], ['34278', 'James Lee', 28.0, ['BEAUTY', 'CLOTHES', 'ELECTRONICS'], [189, 299, 579]]]


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Buen trabajo! Lograste ordenar la lista correctamente.

<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

El método `sort()` en Python funciona correctamente en este caso porque el ID es el primer elemento de cada sublista y, por defecto, `sort()` ordena los elementos de la lista en orden ascendente. Sin embargo, si se quisiera ordenar de otra manera, por ejemplo, en orden descendente, se podría usar `sort(reverse=True)`. Además, si se necesitara ordenar por otro criterio, como la edad o el nombre, se podría utilizar el argumento `key`.

Por ejemplo, así ordenaríamos por la edad de forma ascendente:

	users.sort(key=lambda x: x[2])

</div>

</div>

********Hint********

Puedes utilizar el método `sort()` en la lista de usuarios para ordenarla de forma ascendente.


# Paso 7

Tenemos la información de los hábitos de consumo de nuestros usuarios, incluyendo la cantidad gastada en cada una de sus categorías favoritas. La dirección está interesada en conocer la cantidad total gastada por el usuario.


Calculemos este valor y despleguémoslo.


In [52]:
fav_categories_low = ['electronics', 'sport', 'books']
spendings_per_category = [894, 213, 173]

total_amount = (spendings_per_category[0])+(spendings_per_category[1])+(spendings_per_category[2])

print(total_amount)


1280


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has sumado correctamente los elementos de la lista, obteniendo el resultado esperado de `1280`. 

<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

También podrías utilizar el método sum y obtener la suma de la siguiente manera:

	total_amount = sum(spendings_per_category)

</div>

</div>

********Hint********

¿Cuáles son los tres métodos que se pueden aplicar a una lista para calcular sus valores mínimo, máximo y total?


<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Veo que no has contestado esta pregunta. 📌 Recuerda que es muy importante responder todas las preguntas en los desafíos para demostrar tu comprensión y reforzar tu aprendizaje.

Con respecto a lo que se pregunta, en Python existen tres métodos principales para calcular el mínimo, el máximo y el total de una lista numérica:

- **`min(lista)`** para obtener el valor más bajo.
- **`max(lista)`** para obtener el valor más alto.
- **`sum(lista)`** para calcular la suma total de los valores.

Por ejemplo: 

	valores = [5, 12, 8, 3, 19] 
	print(min(valores)) # 3 
	print(max(valores)) # 19 
	print(sum(valores)) # 47

</div>

# Paso 8

La dirección de la empresa nos pidió pensar en una manera de resumir toda la información de un usuario. Tu objetivo es crear una cadena formateada que utilice información de las variables `user_id`, `user_name` y `user_age`.

Esta es la cadena final que queremos crear: `User 32415 is mike who is 32 years old.` (El usuario 32415 es Mike, quien tiene 32 años).


In [53]:
user_id = '32415'
user_name = ['mike', 'reed']
user_age = 32

user_info = f"User {user_id} is {user_name[0]} who is {user_age} years old."
print(user_info)

User 32415 is mike who is 32 years old.


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has creado la cadena formateada de manera impecable. Estas son muy útiles para integrar valores dinámicos dentro de cadenas de texto de forma legible y clara. Excelente trabajo!

</div>

********Hint********

Para crear una cadena, puedes utilizar el método `format()` o f-string. Para extraer el nombre de la lista `user_name`, puedes utilizar la segmentación.


# Paso 9

La dirección también quiere una forma fácil de conocer la cantidad de clientes con cuyos datos contamos. Tu objetivo es crear una cadena formateada que muestre la cantidad de datos de clientes registrados.

Esta es la cadena final que queremos crear: `Hemos registrado datos de X clientes`.


In [55]:
users = [
    ['32415', ' mike_reed ', 32.0, ['ELECTRONICS', 'SPORT', 'BOOKS'], [894, 213, 173]],
    ['31980', 'kate morgan', 24.0, ['CLOTHES', 'BOOKS'], [439, 390]],
    ['32156', ' john doe ', 37.0, ['ELECTRONICS', 'HOME', 'FOOD'], [459, 120, 99]],
    ['32761', 'SAMANTHA SMITH', 29.0, ['CLOTHES', 'ELECTRONICS', 'BEAUTY'], [299, 679, 85]],
    ['32984', 'David White', 41.0, ['BOOKS', 'HOME', 'SPORT'], [234, 329, 243]],
    ['33001', 'emily brown', 26.0, ['BEAUTY', 'HOME', 'FOOD'], [213, 659, 79]],
    ['33767', ' Maria Garcia', 33.0, ['CLOTHES', 'FOOD', 'BEAUTY'], [499, 189, 63]],
    ['33912', 'JOSE MARTINEZ', 22.0, ['SPORT', 'ELECTRONICS', 'HOME'], [259, 549, 109]],
    ['34009', 'lisa wilson ', 35.0, ['HOME', 'BOOKS', 'CLOTHES'], [329, 189, 329]],
    ['34278', 'James Lee', 28.0, ['BEAUTY', 'CLOTHES', 'ELECTRONICS'], [189, 299, 579]],
]


user_info = f"Hemos registrado datos de {len(users)} clientes"
print(user_info)

Hemos registrado datos de 10 clientes


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Muy bien! Has generado la cadena formateada correctamente, obteniendo `Hemos registrado datos de 10 clientes.`. Esto te permite hacer que los mensajes sean más dinámicos y personalizables.

</div>

********Hint********

Para crear una cadena, puedes utilizar el método `format()` o f-string. Para extraer la cantidad de clientes en la lista, puedes utilizar la función que devuelve la longitud de la lista.


# Paso 10

Apliquemos ahora todos los cambios a la lista de clientes. Para simplificar las cosas, te proporcionaremos una más corta.
Debes:
1. Eliminar todos los espacios iniciales y finales de los nombres, así como cualquier guion bajo.
2. Convertir todas las edades en números enteros.
3. Separar todos los nombres y apellidos en una sublista.

Guarda la lista modificada como una nueva lista llamada `users_clean` y muéstrala en la pantalla.


In [25]:
users = [
    ['32415', ' mike_reed ', 32.0, ['ELECTRONICS', 'SPORT', 'BOOKS'], [894, 213, 173]],
    ['31980', 'kate morgan', 24.0, ['CLOTHES', 'BOOKS'], [439, 390]],
    ['32156', ' john doe ', 37.0, ['ELECTRONICS', 'HOME', 'FOOD'], [459, 120, 99]],
]

users_clean = []


# Procesa al primer usuario
user_name_1 = users[0][1].strip().replace(" ","_")
user_age_1 = int(users[0][2])
user_name_1 = users[0][1].split()
users_clean.append(user_name_1), users_clean.append(user_age_1)

# Procesa al segundo usuario
user_name_2 = users[1][1]
user_age_2 = int(users[1][2])
user_name_2 = users[1][1].split()
users_clean.append(user_name_2), users_clean.append(user_age_2)

# Procesa al tercer usuario
user_name_3 = users[2][1].strip()
user_age_3 = int(users[2][2])
user_name_3 = users[2][1].split()
users_clean.append(user_name_3), users_clean.append(user_age_3)



print(users_clean)


[['mike_reed'], 32, ['kate', 'morgan'], 24, ['john', 'doe'], 37]


<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Excelente trabajo!!

</div>

Solo tienes un pequeño error al momento de agregar los elementos con _append_. 
    
Debería quedarte una lista como la siguiente (presta atención a la estructura de sublistas).

	[['32415', ['mike', 'reed'], 32, ['ELECTRONICS', 'SPORT', 'BOOKS'], [894, 213, 173]], ['31980', ['kate', 'morgan'], 24, ['CLOTHES', 'BOOKS'], [439, 390]], ['32156', ['john', 'doe'], 37, ['ELECTRONICS', 'HOME', 'FOOD'], [459, 120, 99]]]

- Se mantiene la estructura de la lista original.
- Los nombres se disponen en sublistas.
- Las edades en enteros.
- El resto de elementos se mantienen igual que la lista original.
    
    
Para esto, podrías utilizar el _append_ de la siguiente manera (ejemplo con el primer usuario):
    
    users_clean.append([users[0][0], user_name_1, user_age_1, users[0][3], users[0][4]])

</div>

********Hint********

Para procesar a cada usuario, comienza por acceder a los elementos requeridos de la lista de usuarios. Utiliza el método `strip()` para eliminar los espacios iniciales y finales y el método `replace('_',' ')` para reemplazar los guiones bajos por espacios en los nombres. Convierte la edad a un número entero utilizando `int()`. Separa el nombre completo en nombre y apellido utilizando el método `split()`. Por último, `append` (agrega) los datos limpios a la lista `users_clean`.


----------
